rpc/src/services/protocol/mod.rs

```diff
@@ -461,8 +463,19 @@ fn handle_rpc_response(
         ))),
     }
 }
-
-pub(crate) fn call_protocol_rpc(
+/*
+#[cached( name="CALL_PROTOCOL_RPC_CACHE",
+type = "TimedSizedCache<Vec<u8>, serde_json::value::Value>",
+create = "{TimedSizedCache::with_size_and_lifespan(250,20)}",
+convert = "{bincode::serialize(&(chain_param,&chain_id,&block_hash,&rpc_request)).unwrap()}",
+result = true)]
+ */
+#[cached( name="CALL_PROTOCOL_RPC_CACHE",
+type = "TimedSizedCache<(String,ChainId,BlockHash,RpcRequest), serde_json::value::Value>",
+create = "{TimedSizedCache::with_size_and_lifespan(10,20)}",
+convert = "{(chain_param.to_owned(),chain_id.clone(),block_hash.clone(),rpc_request.clone())}",
```

Comment:
@mambisi 
well, this is still no good, rpc_request contains context_path, which could contains `head`,
maybe we should do here a little parsing for that context_path,
like the cache key will be: `chain_id+block_hash+fixed_context_path`

Reply:
ok I get it now

Reply:
I am not sure what the context_path is

Reply:
@mambisi take a look at the `parse_block_hash` function in `rpc/src/helpers.rs` for example, which parses the block fragment (to for example, get a hash from "head").

Reply:
What we need is for example:

- `"/chains/main/blocks/head/helpers/preapply/operations"` -> `"/chains/main/blocks/<some-hash>/helpers/preapply/operations"`
- `"/chains/main/blocks/head~3/helpers/preapply/operations"` -> `"/chains/main/blocks/<some-other-hash>/helpers/preapply/operations"`

etc

Because the meaning of "head" keeps changing all the time, but the meaning of the hash doesn't.

Reply:
Ufff, sorry, I am just making things more confusing. The block hash is already available. Workshop is starting in 3 minutes, we can discuss later. What is important is that the path needs to be normalized so that the `head` part doesn't affect anything.

Reply:
@mambisi so, as @bkontur said, what is needed here is a "fixed" or "normalized" version of the path, because both the chain id and the block has have been obtained already, so what is needed here is some kind of reverse resolution of the URL, maybe hyper already has that?

- `/chains/main/blocks/head/helpers/preapply/operation` -> `/chains/:chain_id/blocks/:block_id/helpers/preapply/operations`
- `/chains/main/blocks/head~3/operation_hashes` -> `/chains/:chain_id/blocks/:block_id/operation_hashes`

etc

@bkontur something I am thinking about is, what about the requests that go through the router? those probably require some pre-parsing too to be correctly cached, right?

Reply:
From looking at https://tezos.gitlab.io/shell/rpc.html and https://tezos.gitlab.io/active/rpc.html#rpc-index, as long as we take care of the `/chains/:chain_id/blocks/:block_id` it seems it should be mostly enough

Reply:
@tizoc @mambisi 
yes, exactly also like Bruno wrote,

- this discussion is just about caching rpc protocol router
- we need to cache just GET request
- all protocol router calls have fixed prefix: `/chains/:chain_id/blocks/:block_id/`  (because just from block we know protocol)
- so cache key should be instead of:
  ```
  convert = "{(chain_param.to_owned(),chain_id.clone(),block_hash.clone(),rpc_request.clone())}"
  ```
 something like this:
  ```
 convert = "{(chain_id.clone(),block_hash.clone(), cache_normalized_url(&rpc_request))}"
  ```
 or maybe just this:
  ```
  convert = "{cache_normalized_url(&rpc_request, &chain_id, &block_hash)}"  // not need to copy chain_id/block_hash
  ```

- that function `cache_normalized_url`  should just replace here `rpc_request.context.path` this two chain_id/block_id
-- maybe just some easy string replacement is enough

@tizoc 
btw. how do we handle rpc mapping in ocaml router for `/chains/main/blocks/head~3/context/constants`?
do you remove everything before `/context/constants`?

Reply:
@bkontur yes, it is ignored, I just use the values sent as part of the request https://gitlab.com/simplestaking/tezos/-/blob/v1.6.0/src/lib_ffi/ffi_rpc_router.ml#L40

Reply:
@bkontur I mean, the structure needs to match, but the values from the fragments are not used for anything.

Reply:
Btw, if every request we send to the OCaml-side contains that prefix already, for the cache key we can just remove it, and use only the remaining part of the path, no need to normalize.